src/CookieRequest.jl

```diff
@@ -15,38 +13,47 @@ function __init__()
     return
 end
 
+export cookielayer
+
 """
-    request(CookieLayer, method, ::URI, headers, body) -> HTTP.Response
+    cookielayer(req) -> HTTP.Response
 
 Add locally stored Cookies to the request headers.
 Store new Cookies found in the response headers.
 """
-abstract type CookieLayer{Next <: Layer} <: Layer{Next} end
-export CookieLayer
-
-function request(::Type{CookieLayer{Next}},
-                 method::String, url::URI, headers, body;
-                 cookies::Union{Bool, Dict{<:AbstractString, <:AbstractString}}=Dict{String, String}(),
-                 cookiejar::Dict{String, Set{Cookie}}=access_threaded(Dict{String, Set{Cookie}}, default_cookiejar),
-                 kw...) where {Next}
-
-    hostcookies = get!(cookiejar, url.host, Set{Cookie}())
-
-    cookiestosend = getcookies(hostcookies, url)
-    if !(cookies isa Bool)
-        for (name, value) in cookies
-            push!(cookiestosend, Cookie(name, value))
+function cookielayer(handler)
+    return function(req::Request; cookies=true, cookiejar::Dict{String, Set{Cookie}}=access_threaded(Dict{String, Set{Cookie}}, default_cookiejar), kw...)
+        if cookies === true || (cookies isa AbstractDict && !isempty(cookies))
+            url = req.url
+            hostcookies = get!(cookiejar, url.host, Set{Cookie}())
+            cookiestosend = getcookies(hostcookies, url)
+            if !(cookies isa Bool)
+                for (name, value) in cookies
+                    push!(cookiestosend, Cookie(name, value))
+                end
+            end
+            if !isempty(cookiestosend)
+                existingcookie = header(req.headers, "Cookie")
```

Comment:
This is net-new code; I realized our test [here](https://github.com/JuliaWeb/HTTP.jl/blob/8743671f8e208b856defb17110ce2cd6705213c4/test/client.jl#L53) was wrong because by default `cookies=false` and so that request wasn't actually _deleting_ any cookies, but just didn't include them at all. This is apparent because it should only be deleting the single `hey` cookie, yet the test was testing that there were _no_ cookies.

The solution here is that we need to track when we automatically include cookies on a request, because in the case of a redirect, the initial response may return a set-cookie that expires or otherwise modifies a cookie we've already added to the request. When redirecting, we re-use the request headers that were set, but for cookies, we want to filter out any cookies that were automatically included on the 1st request and just let the "auto" logic do its thing by including them again _if they're applicable_. This lets cookies that have expired by removed and not sent in subsequent redirect requests.

It's unfortunate it's kind of lost in such a big refactor PR, but oh well.